### ISSUE TITLE
[BUG] `StackingForecaster` seems to execute a senseless algorithm

### ISSUE THREAD

**Issue by fkiraly** [alan-turing-institute/sktime#2668]
The algorithm inside `StackingForecaster` seems strange.

What it does is take the `fh`, and takes the last window of equal length. On it, it fits a regressor of forecasts against actual.

What this results in is, for a `fh` of length `1`, in a fit on a single data point.
That seems completely nonsensical.
There is also no scientific reference cited for the algorithm.

What it should probably be doing instead is using a rolling window splitter, tabulate the forecasts at the same horizon, over train/test pairs, fit a regressor per forecast horizon, and apply that in the forecast.

I am also listed as author, but I only refactored it to the new interface without looking at the algorithm in detail (as I was assuming that the above one is darried out).

FYI @indinewton, have you looked at it? Does the algorithm not look strange to you?

**Comment by indinewton** [comment]
You are right. But I thought maybe someone who proposed this algorithm in past might have reasoned it in terms of code design. Because a lot of design decisions in beginning were affected by how individual contributors thought of using Sktime with whatever data they had. I call it the "curse_of_air_passengers_data" for load_airline() dataset (because then everywhere defaults are like 12 assuming the ts is always monthly and so on). So I think, during that developer assumed that fh will always be 12, so lets stack it using self.fh. 

Anyway that has to be changed somehow! maybe using test_size parameter as is used in AutoEnsembleForecaster instead of salvaging the same fh everywhere. Because for stacking regressor one should be able to use a different fh than the fh passed to the individual forecasters. But we can discuss that in one of friday meetings.

But my doubts go even further. The fit result of both StackingForecaster and AutoEnsembleForecaster is same. The result of predict() method is different for the two. Since I found it strange I created the discussion Q&A, to get some answers from community.  

So until we get concrete answers there, we have also created a user story on our side to check the effectiveness of StackingForecaster against AutoEnsembleForecaster. Our first idea is to check whether for a big fh like > 12, does it achieve comparable performance. 

My expectations are to get answers like:
- either we need StackingForecaster and it makes sense to use it under such and such conditions
- or StackingForecaster is consistently worse than AutoEnsembleForecaster. In that case I would then recommend to remove this class from sktime.

So whatever is the outcome, I will share after June as it is planned for next iteration. Until then lets try to make StackingForecaster sensical code-wise and theory wise.

@danbartl Maybe your thoughts will be useful here? As you have worked with sklearn regressors like GBM in past days. Could you check the StackingForecaster class and share your opinion whether it makes sense to stack forecasters like the way it is proposed in the code? For sensical code, I would ask you to check in stack-fix branch as per https://github.com/alan-turing-institute/sktime/pull/2667

**Comment by fkiraly** [comment]
> For sensical code, I would ask you to check in stack-fix branch as per https://github.com/alan-turing-institute/sktime/pull/2667

#2667 does not actually attempt to fix the content issue, only the interface problem.
The algorithm still seems broken to me there (the intention is unchanged).

**Comment by benHeid** [comment]
Definitely still a bug:

```python
import numpy as np
import pandas as pd
from sklearn.base import BaseEstimator, RegressorMixin
from sktime.forecasting.compose import StackingForecaster
from sktime.forecasting.naive import NaiveForecaster


class RecordingRegressor(RegressorMixin, BaseEstimator):
    """Records the number of samples it is fit on (shared across clones)."""

    calls = []

    def fit(self, X, y):
        RecordingRegressor.calls.append(X.shape[0])
        self._mean = float(np.mean(y))
        return self

    def predict(self, X):
        return np.full(X.shape[0], self._mean)


# small synthetic univariate series, no network access needed
y = pd.Series(np.arange(1.0, 21.0), index=pd.RangeIndex(20))

forecasters = [
    ("naive_last", NaiveForecaster(strategy="last")),
    ("naive_mean", NaiveForecaster(strategy="mean")),
]

sf = StackingForecaster(forecasters=forecasters, regressor=RecordingRegressor())
sf.fit(y, fh=[1])
sf.predict()

assert RecordingRegressor.calls, "Meta-regressor's fit() was never called"
n_samples_used = RecordingRegressor.calls[-1]

print(f"StackingForecaster fit its meta-regressor on {n_samples_used} sample(s) for fh=[1].")

assert n_samples_used > 1, (
    f"StackingForecaster's meta-regressor was fit on only {n_samples_used} data point(s) "
    "for a horizon of length 1. As reported in the bug thread, the current algorithm "
    "windows the training data to a length equal to len(fh), so a single-step forecast "
    "horizon degenerates to fitting a regression model on a single sample -- 'that seems "
    "completely nonsensical.' A fixed implementation (e.g. using a rolling-window CV "
    "splitter, as suggested in the thread) should produce multiple train/test pairs and "
    "therefore fit the meta-regressor on more than one sample even when fh has length 1."
)
```

Triaged by Claude Code
